framework/include/base/ChainControlDataSystem.h

```diff
@@ -0,0 +1,134 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ChainControlData.h"
+#include "MooseError.h"
+#include "MooseUtils.h"
+
+class ChainControl;
+
+/**
+ * System that manages ChainControls.
+ *
+ * To be owned by the MooseApp.
+ */
+class ChainControlDataSystem
+{
+public:
+  ChainControlDataSystem(MooseApp & app);
+
+  /**
+   * Queries if the chain control data of the given name exists.
+   *
+   * @param[in] data_name  Chain control data name
+   * @return True if the chain control data of the given name exists
+   */
+  bool hasChainControlData(const std::string & data_name) const;
+
+  /**
+   * Queries if the chain control data of the given name and type exists.
+   *
+   * @tparam T             Type of the data
+   * @param[in] data_name  Chain control data name
+   * @return True if the chain control data of the given name and type exists
+   */
+  template <typename T>
+  bool hasChainControlDataOfType(const std::string & data_name) const;
+
+  /**
+   * Gets the chain control data of the given name and type and creates if it does not exist.
+   *
+   * If the data exists but of another type, then an error is thrown.
+   *
+   * @tparam T             Type of the data
+   * @param[in] data_name  Chain control data name
+   * @return Reference to the requested chain control data
+   */
+  template <typename T>
+  ChainControlData<T> & getChainControlData(const std::string & data_name);
+
+  /**
+   * Declares chain control data of of the given name and type and creates if it does not exist.
+   *
+   * If the data has already been declared, an error is thrown; otherwise the
+   * data is now set to be declared, and the declaring ChainControl is captured.
+   *
+   * @tparam T                  Type of the data
+   * @param[in] data_name       Chain control data name
+   * @param[in] chain_control   Chain control that is declaring the data
+   * @return Reference to the declared chain control data
+   */
+  template <typename T>
+  ChainControlData<T> & declareChainControlData(const std::string & data_name,
+                                                ChainControl * chain_control);
```

Comment:
why is this a pointer? Is it allowed to be null?

Reply:
Doesn't seem to be any good reason - I'll change it